c8d: Implement prune

vvoland opened the pull request:
- Depends on: https://github.com/moby/moby/pull/45289

Upstreams:
- https://github.com/rumpl/moby/pull/103
- https://github.com/rumpl/moby/pull/20
- https://github.com/rumpl/moby/pull/60

**- What I did**
Implemented prune from fork and adjusted to upstream code.
Also fixed excluding images being used by containers.

Warning: This doesn't handle containers run with `docker run 124c7d2` correctly as it won't exclude their images from being pruned. This doesn't affect the container though, because as long as the snapshot is not deleted, the container will continue to be functional until it's deleted. This may be a correct behavior though?

This is not perfect, but is at least better than not having prune at all.



**- How I did it**

**- How to verify it**
```bash
$ docker run alpine
Unable to find image 'alpine:latest' locally
124c7d270790: Download complete
b5a5b7ce4eab: Download complete
51e60588ff2c: Download complete
c41833b44d91: Download complete
x$ docker run busybox
Unable to find image 'busybox:latest' locally
b5d6fe071263: Download complete
ba7000206594: Download complete
3fbaf71a998b: Download complete
b50100f25006: Download complete
$ docker image prune
WARNING! This will remove all dangling images.
Are you sure you want to continue? [y/N] y
Total reclaimed space: 0B
$ docker ps -a
CONTAINER ID   IMAGE     COMMAND     CREATED          STATUS                      PORTS     NAMES
d8175435dd9b   busybox   "sh"        17 seconds ago   Exited (0) 15 seconds ago             modest_haibt
e735d30840bc   alpine    "/bin/sh"   20 seconds ago   Exited (0) 19 seconds ago             elastic_herschel
$ docker rm d8175435dd9b
d8175435dd9b
$ docker image prune -a
WARNING! This will remove all images without at least one container associated to them.
Are you sure you want to continue? [y/N] y
Deleted Images:
untagged: docker.io/library/busybox:latest
deleted: sha256:b5d6fe0712636ceb7430189de28819e195e8966372edfc2d9409d79402a0dc16
deleted: sha256:ba7000206594c2d72c3ab550453004c0dc50961157e5ebd2fb8ea1890099d02d
deleted: sha256:3fbaf71a998bae6e375be74b999bd418091bf6511e356a129fdc969c4a94a5bc
deleted: sha256:b50100f25006c29bd3a3dd4abacfeb7e9cb61c1a758d07c68fa699a2494fd2df

Total reclaimed space: 2.005MB
$ docker rm e735d30840bc
e735d30840bc
$ docker image prune
WARNING! This will remove all dangling images.
Are you sure you want to continue? [y/N] y
Total reclaimed space: 0B
$ docker image prune -a
WARNING! This will remove all images without at least one container associated to them.
Are you sure you want to continue? [y/N] y
Deleted Images:
untagged: docker.io/library/alpine:latest
deleted: sha256:124c7d2707904eea7431fffe91522a01e5a861a624ee31d03372cc1d138a3126
deleted: sha256:b5a5b7ce4eabc8414bf367761a28f4e8b16952ce5de537c15ed917b71b245f11
deleted: sha256:51e60588ff2cd9f45792b23de89bfface0a7fbd711d17c5f5ce900a4f6b16260
deleted: sha256:c41833b44d910632b415cd89a9cdaa4d62c9725dc56c99a7ddadafd6719960f9

Total reclaimed space: 3.266MB
$ docker images
REPOSITORY   TAG       IMAGE ID   CREATED   SIZE
$ ctr image ls
REF TYPE DIGEST SIZE PLATFORMS LABELS
$ ctr content ls
DIGEST  SIZE    AGE     LABELS
```


**- A picture of a cute animal (not mandatory but encouraged)**

vvoland committed:
c8d: Implement prune

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/list: Implement `until` filter

It's not originally supported by image list, but we need it for `prune`
needs it, so `list` gets it for free.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented:
Also, having an issue with images built by buildkit - it holds the lease and underlying content can't be deleted (although prune correctly deletes the image).

<details>

<summary>SPOILER</summary>

```bash
$ echo 'FROM ubuntu' | docker buildx build -
...
$ docker images
REPOSITORY   TAG       IMAGE ID       CREATED             SIZE
<none>       <none>    3b0d1f17aa98   5 seconds ago       27.4MB

$ docker image prune
WARNING! This will remove all dangling images.
Are you sure you want to continue? [y/N] y
Deleted Images:
untagged: moby-dangling@sha256:0e097f4c8f72f7d497199503e647e4437b8d35249341f2b3733e6c8fdea0e8f6

Total reclaimed space: 0B

$ ctr content ls
DIGEST                                                                  SIZE    AGE             LABELS
sha256:0e097f4c8f72f7d497199503e647e4437b8d35249341f2b3733e6c8fdea0e8f6 855B    7 seconds       containerd.io/gc.ref.content.0=sha256:5c469ba407ba7ee47fdb025546c8490dcca6aa3d2806f51d2c4f183ff72fa2b2,containerd.io/gc.ref.content.1=sha256:78fbaa10733b6fb9b7d2eaafb92cfcd76a5b3ecaf90c5ea3309153b905556bf1
sha256:182b30e356b7623d1e9a65b0c7139321ef929cd2111e4ecfc0230f779e6759c8 4.305kB 5 seconds       -
sha256:2bf0095935ffb29018664cf219d8c1b2c890e3e3c4af89113df23e7330397187 2.314kB 8 seconds       containerd.io/distribution.source.docker.io=library/ubuntu
sha256:380f7223e03db7875bbae70f4c8f463e5523474046e0928bdf84cfaf2ffa2ccd 888B    7 seconds       -
sha256:537da24818633b45fcb65e5285a68c3ec1f3db25f5ae5476a7757bc8dfae92a3 424B    8 seconds       containerd.io/distribution.source.docker.io=library/ubuntu
sha256:5c469ba407ba7ee47fdb025546c8490dcca6aa3d2806f51d2c4f183ff72fa2b2 482B    8 seconds       containerd.io/gc.ref.content.0=sha256:65ed03ad4005007a4116dfb415a623442ee3f61033ef77eb3b44203c437f67b5
sha256:65ed03ad4005007a4116dfb415a623442ee3f61033ef77eb3b44203c437f67b5 1.519kB 8 seconds       containerd.io/gc.ref.snapshot.overlayfs=sha256:874b048c963ab55b06939c39d59303fb975d323822a4ea48a02ac8dc635ea371
sha256:67211c14fa74f070d27cc59d69a7fa9aeff8e28ea118ef3babc295a0428a6d21 1.133kB 8 seconds       containerd.io/distribution.source.docker.io=library/ubuntu
sha256:78fbaa10733b6fb9b7d2eaafb92cfcd76a5b3ecaf90c5ea3309153b905556bf1 565B    7 seconds       containerd.io/gc.ref.content.0=sha256:dbd2f0d7d68abe2ecc4b94d10e98dbbeb35b46f5c778968ec904c65b293cf24a,containerd.io/gc.ref.content.1=sha256:380f7223e03db7875bbae70f4c8f463e5523474046e0928bdf84cfaf2ffa2ccd
sha256:ad18cfdb19dac67bf0072dacea661a817330e5c955d081f4d09914e743ae5d4a 424B    8 seconds       containerd.io/distribution.source.docker.io=library/ubuntu
sha256:b8d7dd79583943e8b6dbc108eb0b99a61d116affe5a377991b4650a9a00565e3 4.062kB 5 seconds       -
sha256:bab8ce5c00ca3ef91e0d3eb4c6e6d6ec7cffa9574c447fd8d54a8d96e7c1c80e 2.316kB 8 seconds       containerd.io/distribution.source.docker.io=library/ubuntu
sha256:cd741b12a7eaa64357041c2d3f4590c898313a7f8f65cd1577594e6ee03a8c38 27.35MB 6 seconds       buildkit.io/blob/annotation.containerd.io/uncompressed=sha256:874b048c963ab55b06939c39d59303fb975d323822a4ea48a02ac8dc635ea371,buildkit.io/blob/mediatype=application/vnd.oci.image.layer.v1.tar+gzip,containerd.io/gc.ref.content.blob-sha256:cd741b12a7eaa64357041c2d3f4590c898313a7f8f65cd1577594e6ee03a8c38=sha256:cd741b12a7eaa64357041c2d3f4590c898313a7f8f65cd1577594e6ee03a8c38
sha256:dbd2f0d7d68abe2ecc4b94d10e98dbbeb35b46f5c778968ec904c65b293cf24a 167B    7 seconds       -

$ ctr leases list
ID                                                                                 CREATED AT           LABELS
615338337-XF4I                                                                     2023-04-04T17:13:41Z buildkit/lease.temporary=2023-04-04T17:13:41.615423462Z,containerd.io/gc.expire=2023-04-04T17:18:41Z
ivypixnysqujps86eydaydj60                                                          2023-04-04T17:13:41Z containerd.io/gc.flat=2023-04-04T17:13:41.589983546Z
mo4mnwri5rfjeaackhza9sceq                                                          2023-04-04T17:13:43Z containerd.io/gc.flat=2023-04-04T17:13:43.510409047Z
mo4mnwri5rfjeaackhza9sceq-variants                                                 2023-04-04T17:13:44Z -
moby-image-sha256:51e60588ff2cd9f45792b23de89bfface0a7fbd711d17c5f5ce900a4f6b16260 2023-04-04T15:24:42Z -
ref_1b21n9vth3px125clb87f8r1m                                                      2023-04-04T17:13:45Z -
ref_2c83yn5ot72le912cx05kxqgm                                                      2023-04-04T16:56:35Z -
ref_oqyrd8dx9g2irmlcsrip1nq14                                                      2023-04-04T16:57:17Z -
ref_rlmbnw2642a4ckpyjpue3wt5d                                                      2023-04-04T16:56:48Z -
ym2i4c18v04x4wsqijzg9goui                                                          2023-04-04T17:13:41Z containerd.io/gc.flat=2023-04-04T17:13:41.589485587Z

$ docker buildx prune -a
WARNING! This will remove all build cache. Are you sure you want to continue? [y/N] y
ID                                              RECLAIMABLE     SIZE            LAST ACCESSED
mo4mnwri5rfjeaackhza9sceq                       true            27.35MB         16 seconds ago
ym2i4c18v04x4wsqijzg9goui*                      true    8.192kB         16 seconds ago
ivypixnysqujps86eydaydj60*                      true    4.096kB         16 seconds ago
Total:  27.36MB

$ ctr content ls
DIGEST                                                                  SIZE    AGE             LABELS
sha256:0e097f4c8f72f7d497199503e647e4437b8d35249341f2b3733e6c8fdea0e8f6 855B    20 seconds      containerd.io/gc.ref.content.0=sha256:5c469ba407ba7ee47fdb025546c8490dcca6aa3d2806f51d2c4f183ff72fa2b2,containerd.io/gc.ref.content.1=sha256:78fbaa10733b6fb9b7d2eaafb92cfcd76a5b3ecaf90c5ea3309153b905556bf1
sha256:182b30e356b7623d1e9a65b0c7139321ef929cd2111e4ecfc0230f779e6759c8 4.305kB 18 seconds      -
sha256:380f7223e03db7875bbae70f4c8f463e5523474046e0928bdf84cfaf2ffa2ccd 888B    20 seconds      -
sha256:5c469ba407ba7ee47fdb025546c8490dcca6aa3d2806f51d2c4f183ff72fa2b2 482B    20 seconds      containerd.io/gc.ref.content.0=sha256:65ed03ad4005007a4116dfb415a623442ee3f61033ef77eb3b44203c437f67b5
sha256:65ed03ad4005007a4116dfb415a623442ee3f61033ef77eb3b44203c437f67b5 1.519kB 20 seconds      containerd.io/gc.ref.snapshot.overlayfs=sha256:874b048c963ab55b06939c39d59303fb975d323822a4ea48a02ac8dc635ea371
sha256:78fbaa10733b6fb9b7d2eaafb92cfcd76a5b3ecaf90c5ea3309153b905556bf1 565B    20 seconds      containerd.io/gc.ref.content.0=sha256:dbd2f0d7d68abe2ecc4b94d10e98dbbeb35b46f5c778968ec904c65b293cf24a,containerd.io/gc.ref.content.1=sha256:380f7223e03db7875bbae70f4c8f463e5523474046e0928bdf84cfaf2ffa2ccd
sha256:b8d7dd79583943e8b6dbc108eb0b99a61d116affe5a377991b4650a9a00565e3 4.062kB 18 seconds      -
sha256:dbd2f0d7d68abe2ecc4b94d10e98dbbeb35b46f5c778968ec904c65b293cf24a 167B    20 seconds      -

$ ctr leases list
ID                                                                                 CREATED AT           LABELS
moby-image-sha256:51e60588ff2cd9f45792b23de89bfface0a7fbd711d17c5f5ce900a4f6b16260 2023-04-04T15:24:42Z -
ref_1b21n9vth3px125clb87f8r1m                                                      2023-04-04T17:13:45Z -
ref_2c83yn5ot72le912cx05kxqgm                                                      2023-04-04T16:56:35Z -
ref_oqyrd8dx9g2irmlcsrip1nq14                                                      2023-04-04T16:57:17Z -
ref_rlmbnw2642a4ckpyjpue3wt5d                                                      2023-04-04T16:56:48Z -
```
</summary>

This is probably related to buildkit 0.11 as it doesn't happen in our fork.

thaJeztah commented:
^^ @tonistiigi @crazy-max

tonistiigi commented:
If these are manifest/attestation/logs blobs then this is expected. They are being kept by the history API and get released with deleting history record or history gc. There shouldn't be any layers/snapshots there though. These also do not show up in `buildx du`.

thaJeztah commented:
I there anything that cleans those up eventually? (or when should they be cleaned up?)

tonistiigi commented:
https://github.com/moby/buildkit/blob/master/solver/llbsolver/history.go#L57-L58 is the default config. This means that nothing will get deleted by gc that is newer than 48h or among the last 50 builds.

vvoland commented:
(It seems I had some leftover state which cluttered the output a bit, updated the result (although I'll check if I can reproduce the lease with ID=image name).
 
Yeah, it seems to be all manifests and attestation related stuff, which is fine I guess, because it's not big. It would be nice if `HistoryQueue` would also have a prune method though.

But, there's still a snapshot of the image rootfs. The snapshot is garbage collected after deleting this lease (`ref_rorjmj0m8y6rwg90304cl5bh7`) so they are related.

vvoland commented:
One minor change I forgot when rebasing: removed `func (i *ImageService) presentChildrenHandler(h containerdimages.HandlerFunc) containerdimages.HandlerFunc` and replaced its usages with `func presentChildrenHandler(store content.Store, h containerdimages.HandlerFunc) containerdimages.HandlerFunc` introduced in #45289 (the behavior is the same).

thaJeztah commented on daemon/containerd/image_prune.go:
```diff
@@ -2,14 +2,151 @@ package containerd
+	// Apply filters
+	for name, img := range imagesToPrune {
+		filteredOut := !filterFunc(img)
+		logrus.WithField("image", name).WithField("filteredOut", filteredOut).Debug("filtering image")
```
For a follow-up; revisit some of these and use `WithFields()` (plural)

thaJeztah commented on daemon/containerd/image_prune.go:
```diff
@@ -2,14 +2,151 @@ package containerd
+		}
+	}
+
+	logrus.WithField("images", imagesToPrune).Debug("pruning")
```
Also for a follow-up; this one seems a bit scary; how large can `imagesToPrune` be? Is that okay do dump in a single log-entry? 😬

thaJeztah commented on daemon/containerd/image_prune.go:
```diff
@@ -2,14 +2,151 @@ package containerd
+			continue
+		}
+
+		report.ImagesDeleted = append(report.ImagesDeleted,
```
For a future release, we really need to start looking at making the API stream the result instead of one giant response with everything that was deleted (and make the CLI show progress while it's happening)

thaJeztah reviewed (approved):
LGTM, thanks!!
